src/hotspot/share/utilities/nonblockingQueue.inline.hpp

```diff
@@ -105,12 +103,23 @@ void NonblockingQueue<T, next_ptr>::append(T& first, T& last) {
   assert(next(last) == NULL, "precondition");
   set_next(last, end_marker());
   T* old_tail = Atomic::xchg(&_tail, &last);
-  if ((old_tail == NULL) ||
+  bool is_old_tail_null = (old_tail == NULL);
+  if (is_old_tail_null ||
       // Try to install first as old_tail's next.
       !is_end(Atomic::cmpxchg(next_ptr(*old_tail), end_marker(), &first))) {
     // Install first as the new head if either
     // (1) the list was empty, or
     // (2) a concurrent try_pop claimed old_tail, so it is no longer in the list.
+    // Note that multiple concurrent push/append operations cannot modify
+    // _head simultaneously, because the Atomic::xchg() above orders these
+    // push/append operations so they perform Atomic::cmpxchg() on different
+    // old_tail. Thus, at most one Atomic::cmpxchg() can fail.
```

Comment:
s/Thus, ... fail./Thus, the cmpxchg can only fail because of a concurrent try_pop./